pruners-ninja@1.0.1 %oneapi, %clang@13, %clang@14: test/ninja_test_util.c:34: multiple definition of `a'

### Steps to reproduce the issue

Reproduced here using:
* `spack@develop` (`08e75f7a3e02ef` from `Fri Aug 12 00:20:47 2022 -0700`)
* `ecpe4s/ubuntu20.04-runner-x86_64-oneapi:2022-07-01`
* `%oneapi@2022.1.0`

Also an issue with:
* `%clang@13.0.1`
* `%clang@14.0.6`

Concrete spec: [pruners-ninja.spec.json.txt](https://github.com/spack/spack/files/9330385/pruners-ninja.spec.json.txt)

Concretization:
```
==> Concretized pruners-ninja
 -   wylz3ag  pruners-ninja@1.0.1%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  tuc6vdd      ^autoconf@2.69%oneapi@2022.1.0 patches=35c4492,7793209,a49dd5b arch=linux-ubuntu20.04-x86_64
[+]  agucw5d          ^m4@1.4.19%oneapi@2022.1.0+sigsegv patches=9dc5fbd,bfdffa7 arch=linux-ubuntu20.04-x86_64
[+]  uli7ace              ^diffutils@3.8%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  sohfg2f                  ^libiconv@1.16%oneapi@2022.1.0 libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  q56bku4              ^libsigsegv@2.13%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  g75tej4          ^perl@5.34.1%oneapi@2022.1.0+cpanm+shared+threads arch=linux-ubuntu20.04-x86_64
[+]  ahtnphg              ^berkeley-db@18.1.40%oneapi@2022.1.0+cxx~docs+stl patches=b231fcc arch=linux-ubuntu20.04-x86_64
[+]  44yzgyw              ^bzip2@1.0.8%oneapi@2022.1.0~debug~pic+shared arch=linux-ubuntu20.04-x86_64
[+]  mnowgf4              ^gdbm@1.19%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  cd2lgld                  ^readline@8.1.2%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  gvjgisz                      ^ncurses@6.2%oneapi@2022.1.0~symlinks+termlib abi=none arch=linux-ubuntu20.04-x86_64
[+]  euvcz47                          ^pkgconf@1.8.0%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  5jz7ihy              ^zlib@1.2.12%oneapi@2022.1.0+optimize+pic+shared patches=0d38234 arch=linux-ubuntu20.04-x86_64
[+]  zl3oyek      ^automake@1.16.5%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  5n46ly5      ^libtool@2.4.7%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  cflgwhp      ^mpich@4.0.2%oneapi@2022.1.0~argobots~cuda+fortran+hwloc+hydra+libxml2+pci~rocm+romio~slurm~two_level_namespace~vci~verbs~wrapperrpath datatype-engine=auto device=ch4 netmod=ofi patches=b4c706e pmi=pmi arch=linux-ubuntu20.04-x86_64
[+]  rltl5ed          ^findutils@4.9.0%oneapi@2022.1.0 patches=440b954 arch=linux-ubuntu20.04-x86_64
[+]  7drxrmz          ^hwloc@2.8.0%oneapi@2022.1.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~oneapi-level-zero~opencl+pci~rocm+shared arch=linux-ubuntu20.04-x86_64
[+]  ky5qiaz              ^libpciaccess@0.16%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  iwdrk44                  ^util-macros@1.19.3%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  go3fobe              ^libxml2@2.9.13%oneapi@2022.1.0~python arch=linux-ubuntu20.04-x86_64
[+]  ppl5hwa                  ^xz@5.2.5%oneapi@2022.1.0+pic libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  mpntixm          ^libfabric@1.14.1%oneapi@2022.1.0~debug~kdreg fabrics=rxm,sockets,tcp,udp arch=linux-ubuntu20.04-x86_64
[+]  dcm6vaw          ^yaksa@0.2%oneapi@2022.1.0~cuda~rocm arch=linux-ubuntu20.04-x86_64
[+]  cg23twc              ^python@3.8.13%oneapi@2022.1.0+bz2+ctypes+dbm~debug+ensurepip+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93,4c24573,f2fd060 arch=linux-ubuntu20.04-x86_64
[+]  ijry5ro                  ^expat@2.4.8%oneapi@2022.1.0+libbsd arch=linux-ubuntu20.04-x86_64
[+]  flwxuwx                      ^libbsd@0.11.5%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  uxmtkav                          ^libmd@1.0.4%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  sf4jy5o                  ^gettext@0.21%oneapi@2022.1.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu20.04-x86_64
[+]  tjngmek                      ^tar@1.34%oneapi@2022.1.0 zip=pigz arch=linux-ubuntu20.04-x86_64
[+]  fisjzhz                          ^pigz@2.7%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  rg3ig4l                          ^zstd@1.5.2%oneapi@2022.1.0+programs compression=none libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  maq3kzw                  ^libffi@3.4.2%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  4455rwb                  ^openssl@1.1.1q%oneapi@2022.1.0~docs~shared certs=mozilla patches=3fdcf2d arch=linux-ubuntu20.04-x86_64
[+]  tlqunsr                      ^ca-certificates-mozilla@2022-07-19%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  7edvlef                  ^sqlite@3.38.5%oneapi@2022.1.0+column_metadata+dynamic_extensions+fts~functions+rtree arch=linux-ubuntu20.04-x86_64
[+]  k3c4qk3                  ^util-linux-uuid@2.37.4%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
```

Install error:
```
==> Installing pruners-ninja-1.0.1-wylz3agaymjpkcqqqek5pqszdd6ut7xf
==> No binary for pruners-ninja-1.0.1-wylz3agaymjpkcqqqek5pqszdd6ut7xf found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/53/53df5c019054b60c68e63d3e249127f1d5f267a70539c8809fb42a8ddbfcb29b.tar.gz
==> No patches needed for pruners-ninja
==> pruners-ninja: Executing phase: 'autoreconf'
==> pruners-ninja: Executing phase: 'configure'
==> pruners-ninja: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'V=1'

6 errors found in build log:
     214    /bin/bash ../libtool  --tag=CC   --mode=link /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-cflgwhpy55pyz2ii3ayp54rxbst3pxwj/bin/
            mpicc  -g -O2   -o ninja_test_matching_race ninja_test_matching_race.o ninja_test_util.o  -lmpi
     215    libtool: link: /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-cflgwhpy55pyz2ii3ayp54rxbst3pxwj/bin/mpicc -g -O2 -o ninja_test_pin
            gpong ninja_test_pingpong.o ninja_test_util.o  -lmpi
     216    mv -f .deps/ninja_test_units.Tpo .deps/ninja_test_units.Po
     217    /bin/bash ../libtool  --tag=CC   --mode=link /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-cflgwhpy55pyz2ii3ayp54rxbst3pxwj/bin/
            mpicc  -g -O2   -o ninja_test_units ninja_test_units.o ninja_test_util.o  -lmpi
     218    libtool: link: /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-cflgwhpy55pyz2ii3ayp54rxbst3pxwj/bin/mpicc -g -O2 -o ninja_test_hyp
            re_parasails ninja_test_hypre_parasails.o ninja_test_util.o  -lmpi
     219    libtool: link: /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-cflgwhpy55pyz2ii3ayp54rxbst3pxwj/bin/mpicc -g -O2 -o ninja_test_mat
            ching_race ninja_test_matching_race.o ninja_test_util.o  -lmpi
  >> 220    /usr/bin/ld: ninja_test_util.o:/tmp/root/spack-stage/spack-stage-pruners-ninja-1.0.1-wylz3agaymjpkcqqqek5pqszdd6ut7xf/spack-src/test/ninja_test_util.c:3
            4: multiple definition of `a'; ninja_test_hypre_parasails.o:/tmp/root/spack-stage/spack-stage-pruners-ninja-1.0.1-wylz3agaymjpkcqqqek5pqszdd6ut7xf/spack
            -src/test/ninja_test_hypre_parasails.c:309: first defined here
     221    libtool: link: /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-cflgwhpy55pyz2ii3ayp54rxbst3pxwj/bin/mpicc -g -O2 -o ninja_test_uni
            ts ninja_test_units.o ninja_test_util.o  -lmpi
  >> 222    clang: error: linker command failed with exit code 1 (use -v to see invocation)
  >> 223    make[3]: *** [Makefile:425: ninja_test_hypre_parasails] Error 1
     224    make[3]: *** Waiting for unfinished jobs....
     225    make[3]: Leaving directory '/tmp/root/spack-stage/spack-stage-pruners-ninja-1.0.1-wylz3agaymjpkcqqqek5pqszdd6ut7xf/spack-src/test'
  >> 226    make[2]: *** [Makefile:485: all-recursive] Error 1
     227    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-pruners-ninja-1.0.1-wylz3agaymjpkcqqqek5pqszdd6ut7xf/spack-src/test'
  >> 228    make[1]: *** [Makefile:422: all-recursive] Error 1
     229    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-pruners-ninja-1.0.1-wylz3agaymjpkcqqqek5pqszdd6ut7xf/spack-src'
  >> 230    make: *** [Makefile:354: all] Error 2
```

### Error message

<details><summary>Error message</summary><pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 0.19.0.dev0 (d735c49b3217598e3a4432f6a6337916f57f730d)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-cascadelake
* **Concretizer:** clingo

### Additional information

[spack-build-out.txt](https://github.com/spack/spack/files/9330395/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/9330396/spack-build-env.txt)

@hainest @lee218llnl @wspear

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/pruners-ninja/package.py
+++ b/var/spack/repos/builtin/packages/pruners-ninja/package.py
@@ -22,3 +22,5 @@ class PrunersNinja(AutotoolsPackage):
     depends_on("autoconf", type="build")
     depends_on("automake", type="build")
     depends_on("libtool", type="build")
+
+    patch("pruners-mutli-def-a-pr3-fix.patch")

--- a/var/spack/repos/builtin/packages/pruners-ninja/pruners-mutli-def-a-pr3-fix.patch
+++ b/var/spack/repos/builtin/packages/pruners-ninja/pruners-mutli-def-a-pr3-fix.patch
@@ -1,0 +1,24 @@
+diff -ruN spack-src/test/ninja_test_hypre_parasails.c spack-src-patched/test/ninja_test_hypre_parasails.c
+--- spack-src/test/ninja_test_hypre_parasails.c	2017-04-25 12:24:43.000000000 -0700
++++ spack-src-patched/test/ninja_test_hypre_parasails.c	2023-06-28 07:41:49.676373145 -0700
+@@ -306,7 +306,7 @@
+ 
+ #define LEN (64)
+ float calibration = 1;
+-int a[LEN];
++static int a[LEN];
+ static void do_noise_work(float msec, int bool)
+ {
+ #ifdef ENABLE_NOISE
+diff -ruN spack-src/test/ninja_test_util.c spack-src-patched/test/ninja_test_util.c
+--- spack-src/test/ninja_test_util.c	2017-03-13 15:36:28.000000000 -0700
++++ spack-src-patched/test/ninja_test_util.c	2023-06-28 07:42:06.444492687 -0700
+@@ -31,7 +31,7 @@
+ 
+ #define LEN (2)
+ int lpusec;
+-int a[LEN];
++static int a[LEN];
+ static void do_noise_work(int loops)
+ { 
+   int i;

